File: 2022/Python/day_2.py
# Me to win
# A Y
# B Z
# C X
#
# them to win
# A Z
# B X
# C Y

import pandas as pd
import numpy as np

data_d2 = pd.read_csv('data/day_2.csv', header = None)
data_d2.columns = ["theirs", "mine"]

combos = np.array([(x, y) for x in ("A", "B", "C") for y in (["X", 1], ["Y", 2], ["Z", 3])], dtype = "object")
combos_df = pd.DataFrame(combos, columns = ['them','me'])
type_scores = pd.DataFrame(combos_df['me'].tolist(),\
     columns = ['me', 'type_score']).join(combos_df["them"])

win = pd.DataFrame(np.array([["A", "Y", 6], ["B", "Z", 6], ["C", "X", 6]]), columns = ["them", "me", "result_score"])
lose = pd.DataFrame(np.array([["A", "Z", 0], ["B", "X", 0], ["C", "Y", 0]]), columns = ["them", "me", "result_score"])
draw = pd.DataFrame(np.array([["A", "X", 3], ["B", "Y", 3], ["C", "Z", 3]]), columns = ["them", "me", "result_score"])
result_scores = pd.concat([win, lose, draw])

scores_joined = type_scores.merge(result_scores) 

File: 2022/R/day_2.R
# Day 1
# What would your total score be if everything goes exactly according to your strategy guide?
# A x rock 1
# B y paper 2
# C z scissors 3

# Me to win
# A Y
# B Z
# C X
#
# them to win
# A Z
# B X
# C Y

# 0 = loss, 3 = draw, 6 = win

library(dplyr)

data_d2 <- read.csv('2022/data/day_2.csv', header = FALSE) %>% 
  rename(them = V1, 
         me = V2)

type_scores <- data.frame("me" = mine,
                          "type_score" = c(1:3))

theirs <- LETTERS[1:3]
mine <- LETTERS[24:26]
combos <- expand.grid("them" = theirs, "me" = mine) %>% 
  mutate(
    result_score = case_when(
      (them == "A" & me == "Y") | (them == "B" & me == "Z") | (them == "C" & me == "X") ~ 6,
      (them == "A" & me == "X") | (them == "B" & me == "Y") | (them == "C" & me == "Z") ~ 3,
      TRUE ~ 0
  )) %>% 
  left_join(type_scores, by = "me") %>% 
  mutate(score = result_score + type_score)

all_games <- data_d2 %>% 
  left_join(combos, by = c("them", "me"))

score <- sum(all_games$score)

